### ISSUE TITLE
Implement positions API for Nicenet

### ISSUE THREAD

**Issue by ashleyvega** [vegaprotocol/specs#100]
In GitLab by @tamlyn10 on Apr 7, 2019, 14:55

This is the task which will come off the design ticket.

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 7, 2019, 14:58

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 7, 2019, 14:59

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 7, 2019, 15:01

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 7, 2019, 15:32

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 7, 2019, 17:08

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 7, 2019, 17:15

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 8, 2019, 21:12

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 12, 2019, 22:26

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Apr 26, 2019, 13:01

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on May 1, 2019, 06:32

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on May 1, 2019, 06:33

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on May 1, 2019, 11:52

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on May 1, 2019, 16:58

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 3, 2019, 01:29

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 3, 2019, 01:29

changed title from **Position{- Engine-}** to **Position{+s+}**

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 3, 2019, 01:35

@tamlyn10 please take a look at the updated description (still a WIP - see below)

I haven’t made any updates to your pseudocode and logic yet but I have separated the Position Engine and Position API logic out after a discussion with @cdm, and based on the fact that the Position Engine within the core itself only needs much more minimal data than the API and that Chris would like us to be explicit about what needs to be in the core (e.g. powers the protocol) vs. what calculation we may need to do to provide the APIs traders want/need. I also created a separate new set of acceptance criteria for the Position Engine. (The logic you described and the existing acceptance criteria sit primarily with the Positions API.)

I think I have kept everything that you did (I also backed up your original description text in case I messed something up), though the API logic still needs work.

I’ve also left some **TODO**s in where I want to double check things with you.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 3, 2019, 01:36

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on May 3, 2019, 12:15

@barnabee thanks for this, reviewing now. 

First thing to note is that the core may need to know the change in closed position.  The reason for this is that in the mark-to-market settlement, when a trader closes out volume, the mark-to-market algorithm has an implementation choice:

**Implementation choice 1.** 
>Using the trade price as the mark price, the mark-to-market settlement is run on all net open positions as calculated PRIOR to the trade occurring. This means that if one (or more) of the traders is closing out volume, they receive the MTM in the same way they normally would. Then once positions are recalculated and they have a new open volume their margin requirement decreases etc. 

>This all sounds good and avoids the need for knowing the change in closed volume... **except** for the potential scenario when a market wants to use a mark price that is not equal to the trade price.  Why might this happen? Well, there are markets that maybe will want a mark price that is the volume weighted price of the last n trades, or wants to ignore trades that are an extreme price (to avoid fat finger mark to markets and stop outs etc.).  We've never really discussed whether our core protocol will own a single method or multiple (and then nominated by market during setup) as prior to us deciding on continuous MTM settlement the mark price wan't core to our protocol I don't think.  If the risk engine needed a mark price I figured that would form part of its parameters and calibration (@davidsiska-vega ?).  However, now it's massively important.  On that note,  I have created a placeholder ticket for calculation of mark price (#106) .  Of course, for nicenet, we can absolutely always use last trade, but the implications for how we design the position engine is worth thinking through in terms of the future.

**Implementation choice 2.** 
>The settlement engine uses the net open positions of the traders as calculated POST the trade occurring.  It has knowledge of any change in closed position after a trade. It includes the closed volume in the relevant 'collect' or 'distribute' call for collateral. If there is a different mark price to trade price then this is captured in the generalised MTM calculations.  


In my discussion yesterday with @EVODelavega, he has currently favoured the second implementation so it would be good to sort this out ASAP for his benefit.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 18:50

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:03

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:03

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:04

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:08

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:12

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:19

@tamlyn10 I’ve updated the Positions API section of the ticket based on my thinking about how this algorithm might be written. Could you take a look? I have gone about it a quite different way but I think it works… hopefully 😬

Also, I think even closed positions should, in our back end code, use negative for short and positive for long volume but would like your opinion on if I’m right or missing something! My thinking:
- It’s consistent! Positions and volume and trades and whatever can be added and subtracted normally and it just works™
- We need to know if a closed position was long or short anyway. API consumers will be interested in this I’m sure.
- It is trivial to display an absolute size number and a long/short label if needed (or not if not) on the front end, and probably better to do that small amount of extra work in the client in this instance rather than allow UI/display requirements to complicate the design.

Thoughts…?

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on May 7, 2019, 19:22

Great - I look forward to digging into this first thing tomorrow morning. Meanwhile, I'm not sure what a signed closed position actually means... once it's closed, it's neither long nor short?!

I guess you could add info about whether it was long then closed (or vice versa) but not sure how really useful that is in the long run.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:23

@tamlyn10 also re: the implementation choice question, if the MTM settlement calculation has the following 4 values in every case, I think it should work without the position engine needing to maintain a closed volume, no?
- current open position size,
- previous open position size,
- current mark price,
- previous mark price

This is assuming: current = after the event being processed, previous = before the event being processed (which will often by the same as the current value)

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on May 7, 2019, 19:25

yeah, re implementation choice question - i think that's right. and the trade info, of course. so that the trade price is the trader's entry price if different to the mark price.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:29

Re: signed closed position, maybe an example would help (we have discussed this before)

```
Buy 50 -> Open pos is +50
Sell 25 -> Open pos is +25, closed pos is +25
Sell 25 -> Open pos is 0, closed pos is +50
```

…easy, but…

```
Buy 50 -> Open pos is +50
Sell 50 -> Open pos is 0, closed pos 1 is +50
Sell 25 -> Open pos is -25, closed pos 1 is +50
Buy 25 -> Open pos is 0, closed pos 1 is *still* 50, closed pos 2 is -25
```

…and…

```
Buy 50 -> Open pos is +50
Sell 75 -> Open pos is -25, closed pos 1 is +50
Buy 25 -> Open pos is 0, closed pos 1 is *still* +50, closed pos 2 is -25
```

i.e. it doesn’t make sense to me to add all closed volume ever to a single closed pos — if I buy and sell 5 things 10,000 times I don’t really have a closed position of ‘size’ 50,000, I have 10,000 closed positions of size 5

Therefore: **we make a new closed position record every time open position reaches or crosses 0.**


Is that reasonable/useful? Seems intuitive and right to me but maybe I am going mad? :upside_down_face:

Note: my API logic and pseudocode currently implement this as described in this comment.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:31

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:38

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 7, 2019, 19:40

Oh, and I guess just to explain the value of negatives in that case, now each closed position was either always long while it was an open/active position or always short, so it makes sense that it would be nice to know that, if you agree with the rest of it of course :smirk_cat:

**Comment by ashleyvega** [comment]
In GitLab by @EVODelavega on May 8, 2019, 15:09

Just my 2 cents: Basically, settlement keeps track of the positions/mark price. Whenever a position changes, or the mark price is updated, it assumes that the values it has are the previous position + old mark price, and the new values are current open position + current mark price. That's enough for it to work out what to do. There's no real need to provide it with 4 values per trader I don't think.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 9, 2019, 02:12

Thinking about it, this could make more sense - if settlement engine stores a last price and a last net position for every trader that it has settled already and the settle function (same both for MTM and at expiry) receives only a current position and price, that does guarantee that the settlement can’t miss a move...

Hmmm 🤔

As long as we have a single "settle" process in each case iaube like that more. Does seem a little safer and can't immediately see any downsides except the cost of storing the values for all positions in the settlement engine persistently rather than only having 2 position values briefly and only for a couple of positions.

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on May 16, 2019, 03:17

Mmmm.. @barnabee  I don't get how the settle function can return the correct cashflows if it doesn't have the previous mark price?

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on May 16, 2019, 18:11

@tamlyn10 The discussion is about whether the position engine stores the previous mark price or is passed it, not whether it’s needed in the calculation.

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 20, 2019, 10:30

marked this issue as related to trading-core#272

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 20, 2019, 12:30

marked this issue as related to trading-core#265

**Comment by ashleyvega** [comment]
In GitLab by @edd on May 31, 2019, 18:16

removed the relation with trading-core#265

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Jun 21, 2019, 12:22

changed title from **{-Positions-}** to **{+Implement positions API for Nicenet+}**

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Jun 21, 2019, 12:22

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Jun 21, 2019, 14:53

assigned to @tamlyn10 and @barnabee

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Sep 20, 2019, 15:38

marked this issue as related to trading-core#457

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Sep 20, 2019, 15:38

I have linked trading-core#457 it will provide a review on the engineering implementation side against the requirements and specification to provide clarity over the state of the system with respect to `nicenet`

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Oct 10, 2019, 17:01

assigned to @cdm

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Nov 26, 2019, 13:25

removed the relation with trading-core#272

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Nov 26, 2019, 13:26

removed the relation with trading-core#457

**Comment by ashleyvega** [comment]
In GitLab by @EVODelavega on Nov 26, 2019, 13:26

marked this issue as related to trading-core#538

**Comment by ashleyvega** [comment]
In GitLab by @EVODelavega on Nov 26, 2019, 13:26

ping: Currently waiting on specs to implement the positions API

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Dec 3, 2019, 12:20

For Nicenet, this is only calculated for positions, rather than on a trade by trade basis. There is a new ticket to implement this #180 on a trade-by-trade basis.
